### PR TITLE
feat(bigquery): export HivePartitioningOptions in load job configurations

### PR DESCRIPTION
**Pull request by shollyman**
Fixes: https://github.com/googleapis/google-cloud-go/issues/3876

### CODE REVIEW

**Review by hongalex — APPROVED** [review]
nit: conventional commits for commit message

feat(bigquery): export ....

**Comment by shollyman** [comment]
> nit: conventional commits for commit message

Weird that the conventional commit bot didn't seem to have issue with it.  Updated.

**Comment by hongalex** [comment]
> > nit: conventional commits for commit message
> 
> Weird that the conventional commit bot didn't seem to have issue with it. Updated.

I think it just checks for the presence of any word before the colon, which the original go commit format satisfies

File: bigquery/load.go [hunk 1]
```diff
@@ -65,6 +65,10 @@ type LoadConfig struct {
 	// For ingestion from datastore backups, ProjectionFields governs which fields
 	// are projected from the backup.  The default behavior projects all fields.
 	ProjectionFields []string
+
+	// HivePartitioningOptions allows use of Hive partitioning based on the
+	// layout of objects in Google Cloud Storage.
```
Comment (tbpg) [inline bigquery/load.go:70]:
```suggestion
	// layout of objects in Cloud Storage.
```

File: bigquery/load.go [hunk 2]
```diff
@@ -83,6 +87,9 @@ func (l *LoadConfig) toBQ() (*bq.JobConfiguration, io.Reader) {
 			ProjectionFields:                   l.ProjectionFields,
 		},
 	}
+	if l.HivePartitioningOptions != nil {
+		config.Load.HivePartitioningOptions = l.HivePartitioningOptions.toBQ()
```
Comment (tbpg) [inline bigquery/load.go:91]: `toBQ` works with `nil` receivers. So, you can inline this like all of the others.

https://github.com/googleapis/google-cloud-go/blob/bc3bf1cdad3aaedc6fdc2987d6cbf8a3481cadeb/bigquery/external.go#L463-L466

I see `nil` is checked for `HivePartitioningOptions.toBQ` in several places. Those can all be simplified (either in this PR or a followup one).

File: bigquery/load.go [hunk 3]
```diff
@@ -111,6 +118,9 @@ func bqToLoadConfig(q *bq.JobConfiguration, c *Client) *LoadConfig {
 		fc = &s.FileConfig
 		lc.Src = s
 	}
+	if q.Load.HivePartitioningOptions != nil {
```
Comment (tbpg) [inline bigquery/load.go:121]: Same comment.

File: bigquery/load_test.go [hunk 4]
```diff
@@ -342,6 +342,31 @@ func TestLoad(t *testing.T) {
...
+				return g
+			}(),
+			config: LoadConfig{
+				HivePartitioningOptions: &HivePartitioningOptions{
+					Mode:                   CustomHivePartitioningMode,
+					SourceURIPrefix:        "source_uri",
+					RequirePartitionFilter: true,
+				},
+			},
+			want: func() *bq.Job {
```
Comment (tbpg) [inline bigquery/load_test.go:359]: :exploding_head: 

This took me a second.

**Review by tbpg — CHANGES_REQUESTED** [review]

File: bigquery/load_test.go (hunk 4 above)
Comment (shollyman) [reply bigquery/load_test.go:359]: You have no idea how much I wish we weren't in the business of mapping back and forth to discovery representations for complex resources like job configs.

File: bigquery/load.go (hunk 3 above)
Comment (shollyman) [reply bigquery/load.go:119]: done

File: bigquery/load.go (hunk 2 above)
Comment (shollyman) [reply bigquery/load.go]: done.

File: bigquery/load.go (hunk 3 above)
Comment (tbpg) [reply bigquery/load.go:119]: Looks like this still needs an update?

**Review by tbpg — CHANGES_REQUESTED** [review]

**Review by tbpg — APPROVED** [review]

**Pull request merged**